README.md
# SierraHoldingParser

[![Build Status](https://travis-ci.com/NYPL/SierraHoldingParser.svg?branch=main)](https://travis-ci.com/NYPL/SierraHoldingParser) [![GitHub version](https://badge.fury.io/gh/nypl%2FsierraHoldingParser.svg)](https://badge.fury.io/gh/nypl%2FsierraHoldingParser) 

This function takes records from the [SierraUpdatePoller](https://github.com/NYPL/sierraUpdatePollerV2/tree/development) and parses them to extract semantic meaning from the holdings fields and to enhance with data from other sources. The resulting record is validated against an Avro schema and passed to the `HoldingPoster` for persistence in the `Holding` database and passing to the `Holding` Kinesis stream for consumption by other functions.

At the moment two enhancement steps will be taken:

- Location codes are looked up and replaced with an object that contains human-readable labels
- Check-In card/box data is fetched from the Sierra database and incorporated with the corresponding holding record

## Requirements

- ruby 2.7
- AWS CLI

## Dependencies

- nypl_ruby_util@0.0.2
- rspec@3.9.0
- mocha@1.11.2

## Environment Variables

- SCHEMA_TYPE: Avro schema to encode parser messages with
- KINESIS_STREAM: Destination stream for parsed messages
- LOG_LEVEL: Standard logging level. Defaults to INFO
- PLATFORM_API_BASE_URL: URL for the NYPL API, used to fetch Avro schema
- LOCATIONS_JSONLD_URL: URL for JSONLD document representing current mapping of location codes to full location objects

## Installation

This function is developed using the AWS SAM framework, [which has installation instructions here](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html)

To install the dependencies for this function, they must be bundled for this framework and should be done with `rake run_bundler`

## Usage

To run the function locally it may be invoked with rake, where EVENT is the name of the specific event in the `events` directory that you would like to invoke the function with:

`rake run_local[FUNCTION]`

## Testing

Testing is provided via `rspec` with `mocha` for stubbing/mocking. The test suite can be invoked with `rake test`


CHANGELOG.md
# CHANGELOG

## 0.0.4 - 2020-10-15
### Fixed
- Use QA location codes in qa/dev environments
- Better handling of date fields (including seasons and month abbreviations)
- Fix bug in logger when no location is attached to record
- Improve handling of missing 853 fields in records
- Lowercase all field names for 853 chronologies

## 0.0.3 - 2020-10-01
### Fixed
- Cleaned up parsing of holding field (863/866) objects

## 0.0.2 - 2020-09-22
### Added
- Added rubocop for linting and linted all files
### Fixed
- Better locations error handling
- Improved parsing of 853/863 fields
- Updated nypl_core_util and loaded deployment helper class from there

## 0.0.1 - 2020-09-17
### Added
- Initial commit containing basic structure and location parser
- Added supporting documentation and unit test suite
- Added integration with check-in card data


spec/spec_helper.rb
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration

ENV['RECORD_TYPE'] = 'testings'
ENV['LOG_LEVEL'] = 'UNKNOWN'
ENV['KINESIS_STREAM'] = 'test_stream'
ENV['LOCATIONS_JSONLD_URL'] = 'test_jsonld_url'

RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :mocha

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/handler_spec.rb
require_relative '../app'
require_relative './spec_helper'


describe 'handler' do
    describe '#init' do
        before(:each) {
            $initialized = false

            @kms_mock = mock()
            @kms_mock.stubs(:decrypt)
            NYPLRubyUtil::KmsClient.stubs(:new).returns(@kms_mock)
            @avro_mock = mock()
            NYPLRubyUtil::NYPLAvro.stubs(:by_name).returns(@avro_mock)
            @kinesis_mock = mock()
            NYPLRubyUtil::KinesisClient.stubs(:new).returns(@kinesis_mock)
            @locations_mock = mock()
            LocationClient.stubs(:new).returns(@locations_mock)
        }

        after(:each) {
            @kms_mock.unstub(:decrypt)
        }

        it "should invoke clients and logger from the ruby utils gem" do
            init

            expect($kms_client).to eq(@kms_mock)
            expect($in_avro_client).to eq(@avro_mock)
            expect($kinesis_client).to eq(@kinesis_mock)
            expect($location_client).to eq(@locations_mock)
            expect($initialized).to eq(true)
        end
     end

    describe '#handle_event' do
        before(:each) {
            @mock_manager = mock()
            RecordManager.stubs(:new).returns(@mock_manager)
            @records = [{ 'id' => 1 }, { 'id' => 2 }, { 'id' => 3 }]
        }
        it "should invoke validate_record and send_record_to_stream for each record" do
            self.stubs(:init).once
            self.stubs(:validate_record).once.with(1).returns(@records[0])
            self.stubs(:validate_record).once.with(2).returns(@records[1])
            self.stubs(:validate_record).once.with(3).returns(@records[2])
            @mock_manager.stubs(:parse_record).times(3)
            self.stubs(:send_record_to_stream).times(3)

            handle_event(event: {"Records" => [1, 2, 3]}, context: {})
        end

        it "should skip sending records to kinesis is avro validation fails" do
            self.stubs(:init).once
            self.stubs(:validate_record).once.with(1).returns(@records[0])
            self.stubs(:validate_record).once.with(2).raises(HoldingParserError.new)
            self.stubs(:validate_record).once.with(3).returns(@records[2])
            @mock_manager.stubs(:parse_record).twice
            self.stubs(:send_record_to_stream).once.with(@records[0])
            self.stubs(:send_record_to_stream).once.with(@records[2])

            handle_event(event: {"Records" => [1, 2, 3]}, context: {})
        end

        it "should continue processing if sending to kinesis raises an error" do
            self.stubs(:init).once
            self.stubs(:validate_record).once.with(1).returns(@records[0])
            self.stubs(:validate_record).once.with(2).returns(@records[1])
            self.stubs(:validate_record).once.with(3).returns(@records[2])
            @mock_manager.stubs(:parse_record).times(3)
            self.stubs(:send_record_to_stream).once.raises(HoldingParserError.new)
            self.stubs(:send_record_to_stream).once.with(@records[1])
            self.stubs(:send_record_to_stream).once.with(@records[2])

            handle_event(event: {"Records" => [1, 2, 3]}, context: {})
        end
    end

    describe '#validate_record' do
        before(:each) {
            @mock_avro = mock()
            $in_avro_client = @mock_avro
        }

        it "should return a decoded record" do
            @mock_avro.stubs(:decode).once.with('test record').returns('decoded record')

            result = validate_record({"kinesis" => {"data" => "test record"}})
            expect(result).to eq('decoded record')
        end

        it "should raise an error if the record is missing the kinesis data object" do
            @mock_avro.stubs(:decode).never

            expect { self.send(:validate_record, {"other" => {"data" => "record"}}) }.to raise_error(HoldingParserError, "Unable to process incoming Kinesis record")
        end

        it "should raise an error if the decode method fails" do
            @mock_avro.stubs(:decode).raises(AvroError.new('testing'))

            expect { self.send(:validate_record, {"kinesis" => {"data" => "record"}}) }.to raise_error(HoldingParserError, "Incoming kinesis record failed Avro decoding")
        end
    end

    describe '#send_record_to_stream' do
        before(:each) {
            @mock_kinesis = mock()
            $kinesis_client = @mock_kinesis
            
            @test_record = { 'id' => 1 }
        }

        it "should encode and send record when successful" do
            @mock_kinesis.stubs(:<<).once.with(@test_record)

            send_record_to_stream(@test_record)
        end

        it "should raise an error and not invoke kinesis if encoding fails" do
            @mock_kinesis.stubs(:<<).once.raises(AvroError.new('test'))

            expect { self.send(:send_record_to_stream, @test_record) }.to raise_error(HoldingParserError, "Unable to encode Avro record for Kinesis")
        end

        it "should raise an error if unable to send record to kinesis" do
            @mock_kinesis.stubs(:<<).once.raises(NYPLError.new('test'))

            expect { self.send(:send_record_to_stream, @test_record) }.to raise_error(HoldingParserError, "Failed to send encoded record to Kinesis stream")
        end
    end
end

spec/parsedfield_spec.rb
require_relative '../lib/field_parser'
require_relative './handler_spec'

describe ParsedField do
    describe :initialize do
        it 'should initialize with h and y fields' do
            test_parser = ParsedField.new('h_field', 'y_field')

            expect(test_parser.instance_variable_get(:@h_field)).to eq('h_field')
            expect(test_parser.instance_variable_get(:@y_field)).to eq('y_field')
            expect(test_parser.string_rep).to eq('')
            expect(ParsedField.class_variable_get(:@@enumeration_codes)).to eq('abcdef')
            expect(ParsedField.class_variable_get(:@@chronology_codes)).to eq('ijkl')
        end
    end

    describe :generate_string_representation do
        before(:each) {
            @test_parser = ParsedField.new('h_field', 'y_field')
        }

        it 'should add enumeration to the string_rep if present' do
            @test_parser.stubs(:_generate_enumeration).once.returns(['test enum', nil])
            @test_parser.stubs(:_generate_chronology).once.returns([nil, nil])

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('test enum')
        end

        it 'should add enumeration and chronology (in parens) to the string_rep if both present' do
            @test_parser.stubs(:_generate_enumeration).once.returns(['test enum', nil])
            @test_parser.stubs(:_generate_chronology).once.returns(['test chron', nil])

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('test enum (test chron)')
        end

        it 'should add chronology (sans parens) to the string_rep if no enumeration present' do
            @test_parser.stubs(:_generate_enumeration).once.returns([nil, nil])
            @test_parser.stubs(:_generate_chronology).once.returns(['test chron', nil])

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('test chron')
        end

        it 'should add a hyphen to the end of the string if continuing is present' do
            @test_parser.stubs(:_generate_enumeration).once.returns(['test enum', nil])
            @test_parser.stubs(:_generate_chronology).once.returns(['test chron', nil])
            @test_parser.instance_variable_set(:@continuing, true)

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('test enum (test chron)-')
        end

        it 'should create a hypher separated range for start and end enumerations' do
            @test_parser.stubs(:_generate_enumeration).once.returns(['test enum1', 'test enum2'])
            @test_parser.stubs(:_generate_chronology).once.returns([nil, nil])

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('test enum1 - test enum2')
        end

        it 'should create a hypher separated range for start and end chronologies' do
            @test_parser.stubs(:_generate_enumeration).once.returns([nil, nil])
            @test_parser.stubs(:_generate_chronology).once.returns(['chron1', 'chron2'])

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('chron1 - chron2')
        end

        it 'should create a hypher separated range for start and end chronologies and enumerations' do
            @test_parser.stubs(:_generate_enumeration).once.returns(['enum1', 'enum2'])
            @test_parser.stubs(:_generate_chronology).once.returns(['chron1', 'chron2'])

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('enum1 (chron1) - enum2 (chron2)')
        end

        it 'should create a hypher separated range for a single start chronology and start and end enumerations' do
            @test_parser.stubs(:_generate_enumeration).once.returns(['enum1', 'enum2'])
            @test_parser.stubs(:_generate_chronology).once.returns(['chron1', nil])

            @test_parser.generate_string_representation

            expect(@test_parser.string_rep).to eq('enum1 (chron1) - enum2 (chron1)')
        end
    end

    describe :_generate_enumeration do
        it 'should return comma delimited string of subfields in @@enumeration_codes' do
            test_parser = ParsedField.new(
                { 'a' => '1', 'c' => '3' },
                { 'a' => 'v.', 'b' => 'ser.', 'c' => 'iss.' }
            )

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq('v. 1, iss. 3')
            expect(out_str[1]).to eq(nil)
        end

        it 'should return an empty string if there are no enumeration subfields in the h record' do
            test_parser = ParsedField.new({ 'i' => '1999' }, { 'a' => 'v.' })

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq(nil)
        end

        it 'should return an empty string if there are no subfields in the h record' do
            test_parser = ParsedField.new({}, { 'a' => 'v.' })

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq(nil)
        end

        it 'should remove any subfields if the h record is an empty string' do
            test_parser = ParsedField.new(
                { 'a' => '1', 'b' => '', 'c' => '3' },
                { 'a' => 'v.', 'b' => 'ser.', 'c' => 'i.' }
            )

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq('v. 1, i. 3')
        end

        it 'should combine values with colons if no field names are provided' do
            test_parser = ParsedField.new(
                { 'a' => '1', 'b' => '', 'c' => '3' },
                { 'a' => '', 'b' => '', 'c' => '' }
            )

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq('1:3')
        end

        it 'should combine values with colons and commas if values are mixed' do
            test_parser = ParsedField.new(
                { 'a' => '1', 'b' => '2', 'c' => '3' },
                { 'a' => 'ser.', 'b' => 'vol.', 'c' => '' }
            )

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq('ser. 1, vol. 2:3')
        end

        it 'should set continuing if a hyphen is present in a value' do
            test_parser = ParsedField.new(
                { 'a' => '1-', 'b' => '2', 'c' => '3' },
                { 'a' => 'ser.', 'b' => 'vol.', 'c' => '' }
            )

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq('ser. 1, vol. 2:3')
            expect(out_str[1]).to eq(nil)
            expect(test_parser.instance_variable_get(:@continuing)).to eq(true)
        end

        it 'should create a range of values if for hyphen separated fields' do
            test_parser = ParsedField.new(
                { 'a' => '1-3' }, { 'a' => 'no.' }
            )

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq('no. 1')
            expect(out_str[1]).to eq('no. 3')
        end

        it 'should create a range of combined values for multiple hyphen separeted fields' do
            test_parser = ParsedField.new(
                { 'a' => '1-3', 'b' => '20-40' }, { 'a' => 'vol.', 'b' => '' }
            )

            out_str = test_parser.send(:_generate_enumeration)

            expect(out_str[0]).to eq('vol. 1:20')
            expect(out_str[1]).to eq('vol. 3:40')
        end
    end

    describe :_generate_chronology do
        it 'should pass key/value pairs to DateComponent and get string from created object' do
            test_parser = ParsedField.new(
                { 'i' => '1999', 'j' => 'Mar' },
                { 'i' => 'year', 'j' => '(month)' }
            )

            mock_date_comp = mock
            ParsedField::DateComponent.stubs(:new).once.returns(mock_date_comp)
            mock_date_comp.stubs(:set_field).once.with('year', '1999')
            mock_date_comp.stubs(:set_field).once.with('month', 'Mar')
            mock_date_comp.stubs(:create_str).once.returns('1999-Mar')

            out_str = test_parser.send(:_generate_chronology)

            expect(out_str).to eq('1999-Mar')
        end

        it 'should pass no fields to DateComponent if no matching subfields exist and return empty string' do
            test_parser = ParsedField.new(
                { 'a' => '1', 'c' => '3' },
                { 'i' => 'year', 'j' => '(month)' }
            )

            mock_date_comp = mock
            ParsedField::DateComponent.stubs(:new).once.returns(mock_date_comp)
            mock_date_comp.stubs(:set_field).never
            mock_date_comp.stubs(:create_str).once.returns('')

            out_str = test_parser.send(:_generate_chronology)

            expect(out_str).to eq('')
        end

        it 'should omit any fields that are only an empty string' do
            test_parser = ParsedField.new(
                { 'i' => '1999', 'j' => '' },
                { 'i' => 'year', 'j' => 'month' }
            )

            mock_date_comp = mock
            ParsedField::DateComponent.stubs(:new).once.returns(mock_date_comp)
            mock_date_comp.stubs(:set_field).once.with('year', '1999')
            mock_date_comp.stubs(:set_field).never
            mock_date_comp.stubs(:create_str).once.returns('1999')

            out_str = test_parser.send(:_generate_chronology)

            expect(out_str).to eq('1999')
        end

        it 'should handle delimiter fields with abreviations' do
            test_parser = ParsedField.new(
                { 'i' => '1999', 'j' => '09', 'k' => '09' },
                { 'i' => '(yr.)', 'j' => 'mo.', 'k' => 'da' }
            )

            mock_date_comp = mock
            ParsedField::DateComponent.stubs(:new).once.returns(mock_date_comp)
            mock_date_comp.stubs(:set_field).once.with('year', '1999')
            mock_date_comp.stubs(:set_field).once.with('month', '09')
            mock_date_comp.stubs(:set_field).once.with('day', '09')
            mock_date_comp.stubs(:create_str).once.returns('1999-09-09')

            out_str = test_parser.send(:_generate_chronology)

            expect(out_str).to eq('1999-09-09')
        end

        it 'should set continuing to true if a value ends with a hyphen' do
            test_parser = ParsedField.new(
                { 'i' => '1999', 'j' => '23-' },
                { 'i' => 'year', 'j' => '()' }
            )

            mock_date_comp = mock
            ParsedField::DateComponent.stubs(:new).once.returns(mock_date_comp)
            mock_date_comp.stubs(:set_field).once.with('year', '1999')
            mock_date_comp.stubs(:set_field).once.with(nil, '23-')
            mock_date_comp.stubs(:set_field).never
            mock_date_comp.stubs(:create_str).once.returns('1999-23')

            out_str = test_parser.send(:_generate_chronology)

            expect(out_str).to eq('1999-23')
            expect(test_parser.instance_variable_get(:@continuing)).to eq(true)
        end
    end

    describe :_standardize_date_definition_field do
        it 'should return the full name of a date field for an abbreviation' do
            test_parser = ParsedField.new({}, {})

            out_field = test_parser.send(:_standardize_date_definition_field, '(yr.)')

            expect(out_field).to eq('year')
        end

        it 'should return nil if an empty set of parens is provided' do
            test_parser = ParsedField.new({}, {})

            out_field = test_parser.send(:_standardize_date_definition_field, '()')

            expect(out_field).to eq(nil)
        end

        it 'should raise an error if the date field is not recognized' do
            test_parser = ParsedField.new({}, {})

            expect {
                test_parser.send(:_standardize_date_definition_field, '(smthg.)')
            }.to raise_error(FieldParserError, 'Unable to identify field (smthg.) for chronology')
        end

        it 'should handle cases where fields contain upper case characters' do
            test_parser = ParsedField.new({}, {})

            out_field = test_parser.send(:_standardize_date_definition_field, '(Season)')

            expect(out_field).to eq('season')
        end
    end

    describe :_empty_field_check do
        it 'should return true if the field is not an empty string' do
            test_parser = ParsedField.new({}, {})

            expect(test_parser.send(:_empty_field_check, 'test')).to eq(true)
        end

        it 'should return false if the field is an empty string' do
            test_parser = ParsedField.new({}, {})

            expect(test_parser.send(:_empty_field_check, '   ')).to eq(nil)
        end
    end
end


app.rb
require 'nypl_ruby_util'

require_relative 'lib/location_client'
require_relative 'lib/record_manager'

def init
    return if $initialized

    $logger = NYPLRubyUtil::NyplLogFormatter.new($stdout, level: ENV['LOG_LEVEL'])
    $kms_client = NYPLRubyUtil::KmsClient.new
    $in_avro_client = NYPLRubyUtil::NYPLAvro.by_name(ENV['IN_SCHEMA_TYPE'])
    $kinesis_client = NYPLRubyUtil::KinesisClient.new({
        schema_string: ENV['OUT_SCHEMA_TYPE'],
        stream_name: ENV['KINESIS_STREAM'],
        partition_key: 'id'
    })
    $location_client = LocationClient.new

    $logger.debug 'Initialized function'
    $initialized = true
end

# rubocop:disable Lint/UnusedMethodArgument
def handle_event(event:, context:)
    init

    $logger.info 'Beginning processing of record batch'

    event['Records'].each do |record|
        $logger.debug 'Processing record from Kinesis'
        begin
            valid_record = validate_record record
        rescue HoldingParserError => e
            $logger.error("Unable to validate record with reason: #{e.message}")
            next
        end

        record_manager = RecordManager.new valid_record
        record_manager.parse_record

        begin
            send_record_to_stream valid_record
        rescue HoldingParserError => e
            $logger.error("Unable to send record to kinesis with reason: #{e.message}")
            next
        end
        $logger.debug "Processed and sent record # #{valid_record['id']} to kinesis"
    end

    $logger.info 'Processing Complete'
end
# rubocop:enable Lint/UnusedMethodArgument

def validate_record(record)
    $logger.debug 'Validating Record'
    begin
        avro_data = record['kinesis']['data']
    rescue KeyError, NoMethodError => e
        $logger.error "Missing field in Kinesis message, unable to process #{e.message}"
        raise HoldingParserError, 'Unable to process incoming Kinesis record'
    end

    begin
        decoded = $in_avro_client.decode avro_data
        $logger.debug 'Decoded bib', decoded
    rescue AvroError => e
        $logger.error "Record failed Avro decoding for reason: #{e.message}"
        raise HoldingParserError, 'Incoming kinesis record failed Avro decoding'
    end

    decoded
end

def send_record_to_stream(record)
    $kinesis_client << record
    $logger.info "Sent record to kinesis stream record ##{record['id']}"
rescue AvroError => e
    $logger.warn "Record (id# #{record['id']} failed avro validation", { status: e.message }
    raise HoldingParserError, 'Unable to encode Avro record for Kinesis'
rescue NYPLError => e
    $logger.warn "Record (id# #{record['id']} failed to write to kinesis", { status: e.message }
    raise HoldingParserError, 'Failed to send encoded record to Kinesis stream'
end

class HoldingParserError < StandardError; end


lib/location_client.rb
require 'json'
require 'net/http'
require 'uri'

# Class for retrieving location labels from NYPL-Core and adding them to fetched location codes
class LocationClient
    def initialize
        @locations = _load_locations
    end

    def lookup_code(code)
        location = nil
        unless code.strip == 'none'
            code_record = @locations[code.strip]

            unless code_record && code_record['code'] && code_record['label']
                raise NYPLLocationError, "Missing or incomplete code_record: #{code_record} for location code: #{code}"
            end

            location = { 'code' => code_record['code'], 'label' => code_record['label'] }
        end

        location
    end

    private

    def _load_locations
        begin
            location_jsonld_uri = URI(ENV['LOCATIONS_JSONLD_URL'])
            response = Net::HTTP.get_response(location_jsonld_uri)
        rescue StandardError => e
            $logger.error "Unable to fetch locations JSONLD due to #{e.message}"
            raise NYPLLocationError, 'Unable to load NYPL locations data'
        end

        unless response.code.to_i == 200
            $logger.error 'Received non-200 status from locations JSONLD object'
            $logger.debug 'NYPL Location error', { error: response.body }
            raise NYPLLocationError, "NYPL location fetch returned #{response.code}"
        end

        locations_object = JSON.parse(response.body)
        $logger.debug 'Loaded NYPL locations object'

        locations_object
    end
end

class NYPLLocationError < StandardError; end


lib/record_manager.rb
require_relative './field_parser'

# Class that performs the main transformation functions, including
# - Fetching Check-In Card data
# - Fetching location labels
# - Parsing holding objects
class RecordManager
    @@default_y_fields = {
        'a' => '',
        'b' => '',
        'c' => '',
        'd' => '',
        'e' => '',
        'f' => '',
        'i' => 'year',
        'j' => 'month',
        'k' => 'day',
        'l' => ''
    }

    def initialize(record)
        @record = record
    end

    def self.default_y_fields
        @@default_y_fields
    end

    def parse_record
        $logger.info "Parsing record # #{@record['id']}"
        _parse_location
        _parse_holdings
    end

    private

    def _parse_location
        $logger.debug 'Adding location object to record'

        location_object = $location_client.lookup_code @record['fixedFields']['40']['value']
        $logger.debug 'Retrieved location object', location_object || { message: 'No location specified' }

        @record['location'] = location_object
    end

    def _parse_holdings
        $logger.debug 'Setting parsed holdings field'
        @record['holdings'] = []

        # Extract all holdings strings into the holdings field
        _get_h_fields

        # Incorporate Check-In Card data to this object
        _get_check_in_card
    end

    def _get_check_in_card
        $logger.debug "Fetching check-in card data for holding ##{@record['id']}"

        begin
            check_in_card_uri = URI(
                "#{ENV['PLATFORM_API_BASE_URL']}holdings/check-in-cards?holding_id=#{@record['id']}"
            )
            response = Net::HTTP.get_response(check_in_card_uri)
        rescue StandardError => e
            $logger.error 'Failed to load check-in card data', { status: e.message }
            raise RecordError, 'Could not load check-in-card data'
        end

        # Confirm that a valid response was received
        unless response.code.to_i == 200
            $logger.error "Unable to load check-in card data with status #{response.code}", { status: response.body }
            raise RecordError, "Unable to load check-in card data with status #{response.code}"
        end

        # Parse response into an object
        check_in_data = JSON.parse(response.body)
        $logger.debug check_in_data
        @record['checkInCards'] = check_in_data
    end

    def _get_h_fields
        $logger.debug 'Getting values for all h values'

        y_853_fields = @record['varFields'].filter { |f| f['marcTag'] == '853' && f['fieldTag'] == 'y' }

        h_866_fields, h_legacy_fields, h_863_fields = _load_h_fields_by_type

        @record['holdings'].push(*_create_holding_obj(h_866_fields.map { |h| h['subfields'][0]['content'] }))
        @record['holdings'].push(*_create_holding_obj(h_legacy_fields.map { |h| h['content'] }))
        @record['holdings'].push(*_create_holding_obj(_parse_853_863_fields(y_853_fields, h_863_fields)))

        $logger.debug @record['holdings']
    end

    def _load_h_fields_by_type
        all_h_fields = @record['varFields'].filter { |f| f['fieldTag'] == 'h' }

        h_866_fields = all_h_fields.filter { |f| f['marcTag'] == '866' }
        h_legacy_fields = all_h_fields.filter { |f| !f['content'].nil? }
        h_863_fields = all_h_fields.filter { |f| f['marcTag'] == '863' }

        [h_866_fields, h_legacy_fields, h_863_fields]
    end

    def _create_holding_obj(holding_arr)
        holding_arr.map do |h|
            {
                'holding_string' => h,
                'holding_ranges' => [],
                'index' => false,
                'incomplete' => false,
                'negation' => false
            }
        end
    end

    def _parse_853_863_fields(y_fields, h_fields)
        y_map = y_fields.empty? ? { '1' => @@default_y_fields } : _transform_field_array_to_hash(y_fields)
        h_map = _transform_field_array_to_hash h_fields

        h_y_crosswalk = y_map.keys.map { |k| [k, []] }.to_h

        h_map.each do |k, v|
            y_match, position = k.split('.')
            h_y_crosswalk[y_match][position.to_i - 1] = v
        end

        out_strings = []
        h_y_crosswalk.each do |k, v|
            out_strings << v.map do |h|
                next if h.nil?

                field_parser = ParsedField.new(h, y_map[k])
                field_parser.generate_string_representation

                field_parser.string_rep
            end.compact.join('; ')
        end

        out_strings
    end

    def _transform_field_array_to_hash(fields)
        fields.map do |f|
            sub_map = f['subfields'].map { |s| [s['tag'], s['content']] }.to_h

            [sub_map['8'], sub_map.keep_if { |k, _v| k != '8' }]
        end.to_h
    end
end

class RecordError < StandardError; end
